Build instructions focused on stable Server/CLI builds, Test instructions separated (#375)
ydb-platform/ydb: BUILD.md

 - libidn11-dev
 - ninja 1.10+

> We run multiple clang instances in parallel to speed up the process by default. Each instance of clang may use up to 1GB of RAM, and linking the binary may use up to 16GB of RAM, please make sure your build host has enough resources.

Author: Let's add it later. I did not change this text with this PR.

```bash
mkdir ~/ydbwork && cd ~/ydbwork
mkdir build
```

> ## Install ccache

Author: If you're talking about ccache, then the idea was to make the instruction less variable. Those who knows what ccache is will understand that it's optional. Those who do not know -- let them install it.

1. Install `ccache` into `/usr/local/bin/` (We are using version `4.8.1`, you can use any version greater than `4.7`)
    ```bash
    (V=4.8.1; curl -L https://github.com/ccache/ccache/releases/download/v${V}/ccache-${V}-linux-x86_64.tar.xz | \
     sudo tar -xJ -C /usr/local/bin/ --strip-components=1 --no-same-owner ccache-${V}-linux-x86_64/ccache)
    ```

2. Configure `ccache` to use remote storage of YDB build artifacts, to speed up first build time:
    ```bash
    ccache -o remote_storage="http://cachesrv.ydb.tech:8080|read-only|layout=bazel"
    ccache -o sloppiness=locale 
    ccache -o base_dir=~/ydbwork/
    ```

If you use a non-default work directory, correct the `base_dir` ccache option to match it.

## Clone the ydb repository.

```bash

> git clone https://github.com/ydb-platform/ydb.git

Author: So that GitHub authentication is not needed. Instructions on how to develop (make changes) will be in a separate document, and there will be git protocol there.

At the moment of this documentation update, the latest stable release was [YDB Server 23.2.12](https://github.com/ydb-platform/ydb/releases/tag/23.2.12) with a tag '23.2.12', so the command looks like that:

```bash
git checkout 23.2.12
```

### Check out the latest stable YDB CLI version for build

To build a latest stable version of a YDB CLI, check out the latest stable CLI tag from the repository. To do so, visit the https://github.com/ydb-platform/ydb/releases page, scroll down to the top-most 'YDB CLI' release, and use the provided 'tag' value in the `git checkout <tag>` command.

At the moment of this documentation update, the latest YDB CLI release was [YDB CLI 2.5.0](https://github.com/ydb-platform/ydb/releases/tag/CLI_2.5.0) with a tag 'CLI_2.5.0', so the command looks like that:

```bash
git checkout CLI_2.5.0
```

## Configure

1. Change Conan's home folder to the build folder for better remote cache hit 
    ```bash

>     export CONAN_USER_HOME=~/ydbwork/build

Author: Maybe, but that's like it was before this change.

    ```

2. Generate build configuration using `ccache`
    ```bash
    cd build

>     cmake -G Ninja -DCMAKE_BUILD_TYPE=Release \

Reviewer: It probably deserves explaining what kind of build types are available and what are the differences.

Author: Not here, I will make a separate doc for development, reference it here, and describe it there.

    -DCCACHE_PATH=/usr/local/bin/ccache \
    -DCMAKE_TOOLCHAIN_FILE=../ydb/clang.toolchain \
    ../ydb  
    ```

## Build 

### Build YDB Server

To build YDB Server run:
```bash
ninja ydb/apps/ydbd/all
```

A YDB server binary can be found at:
```
ydb/apps/ydbd/ydbd
```

### Build YDB CLI

> To build YDB CLI (Command Line Interface utility) run:

Reviewer: it is better to explain what CLI stands for in the beginning when we first mention it, not here almost at the end

Author: It is actually there, in the first sentence.

Reviewer: Then this is a duplicate explantation and thus not needed?

ydb/apps/ydb/ydb
```

> ## Run tests

Reviewer: One more section is really missing: how to just run the locally built server and connect to it with CLI, without any tests.

Author: Let's add it later.

### Build all executable artifacts

To run tests, you first to build all binary artifacts (tools, test executables, server and CLI, etc.), running `ninja` without parameters:
```bash
ninja
```

### Run unit tests

To run tests execute:
```bash
ctest
```

### Functional tests

Run test suites via pytest according to this [instruction](ydb/tests/functional/README.md).

## Run YDB CLI tests only

> When changing YDB CLI code, you may choose building and running only YDB CLI tests, which is faster than running a full YDB testpack.

Reviewer: ```suggestion
You may choose building and running only YDB CLI tests when changing the YDB CLI code, as it is faster than running the full YDB test suite.
```

To build YDB CLI binary and YDB CLI unit tests binaries run:
```bash
ninja ydb/apps/ydb/all
ninja ydb/public/lib/ydb_cli/all
```

To run YDB CLI unit tests execute:
```bash
cd ydb/public/lib/ydb_cli/
ctest
```

> To run YDB CLI functional tests you need a compiled YDB Server binary (ydbd) located at ~/ydbwork/ydb/apps/ydbd/ydbd. You may compile it as described in the "Build YDB Server" above, or download a precompiled binary from the YDB [Downloads](https://ydb.tech/en/docs/downloads/#ydb-server) page.

Reviewer: ```suggestion
To run YDB CLI functional tests you need a compiled YDB Server binary (ydbd) located at ~/ydbwork/ydb/apps/ydbd/ydbd. You may compile it as described in the [Build YDB Server](#build-ydb-server) above, or download a precompiled binary from the YDB [Downloads](https://ydb.tech/en/docs/downloads/#ydb-server) page.
```
